TrackletAlgorithm/MatchProcessor.h

```diff
@@ -386,40 +1000,78 @@ void MatchCalculator(BXType bx,
+  bool isProjDisk = proj_seed >= TF::D1;
 
   // Calculate residuals
   // Get phi and z correction
   ap_int<22> full_phi_corr = stub_r * proj_phid; // full corr has enough bits for full multiplication
   ap_int<18> full_z_corr   = stub_r * proj_zd;   // full corr has enough bits for full multiplication
+  ap_int<18> full_r_corr   = stub_z * proj_zd;   // full corr has enough bits for full multiplication
   ap_int<11> phi_corr      = full_phi_corr >> kPhi_corr_shift;                        // only keep needed bits
+  //ap_uint<3> shifttmp = 6;
+  constexpr auto shifttmp = (LAYER < trklet::N_LAYER) ? log2barrel : log2disk;
+  if(isDisk && isPSStub)
+    phi_corr = (stub_ps_z * proj_phid) >> shifttmp;
+  else if(isDisk && !isPSStub)
+    phi_corr = (stub_2s_z * proj_phid) >> shifttmp;
   ap_int<12> z_corr        = (full_z_corr + (1<<(kZ_corr_shift-1))) >> kZ_corr_shift; // only keep needed bits
+  ap_int<12> r_corr        = full_r_corr >> kr_corr_shift; // only keep needed bits
    
   // Apply the corrections
   const int kProj_phi_len = AllProjection<APTYPE>::kAProjPhiSize + 1;
-  ap_int<kProj_phi_len> proj_phi_corr = proj_phi + phi_corr;  // original proj phi plus phi correction
+  ap_int<kProj_phi_len> proj_phi_corr = proj_phi + phi_corr;  // original proj phi plus phi correction iphi in emulation
   ap_int<13> proj_z_corr   = proj_z + z_corr;      // original proj z plus z correction
 
   // Get phi and z difference between the projection and stub
-  ap_int<10> delta_z         = stub_z - proj_z_corr;
+  ap_int<12> delta_z        = stub_z - proj_z_corr;
   ap_int<14> delta_z_fact   = delta_z * kFact;
-  ap_int<18> stub_phi_long  = stub_phi;         // make longer to allow for shifting
+  const ap_int<18> &stub_phi_long  = stub_phi;         // make longer to allow for shifting
   const ap_int<18> &proj_phi_long  = proj_phi_corr;    // make longer to allow for shifting
-  ap_int<18> shiftstubphi   = stub_phi_long << kPhi0_shift;                        // shift
-  ap_int<18> shiftprojphi   = proj_phi_long << (kShift_phi0bit - 1 + kPhi0_shift); // shift
-  ap_int<17> delta_phi      = shiftstubphi - shiftprojphi;
-  ap_uint<17> abs_delta_phi = iabs<17>( delta_phi );    // absolute value of delta phi
+  ap_int<18> shiftstubphi   = kPhi0_shift > 0 ? stub_phi_long << kPhi0_shift : stub_phi_long;                        // shift
+  if(isDisk && isPSStub)
+    shiftstubphi = kPhi0_shift > 0 ? stub_ps_phi << kPhi0_shift : stub_ps_phi;                        // shift
+  else if(isDisk && !isPSStub) {
+    shiftstubphi = kPhi0_shift > 0 ? stub_2s_phi << kPhi0_shift : stub_2s_phi;                        // shift
+  }
+  ap_int<18> shiftprojphi = proj_phi_long << (kShift_phi0bit - 1 + kPhi0_shift); // shift
+  constexpr int dphibit = 20;
+  ap_int<dphibit> delta_phi      = shiftstubphi - shiftprojphi;
+  ap_uint<3> shiftprojz     = 7;
+  ap_int<7> proj_r_corr    = (stub_z * proj_zd) >> shiftprojz;
```

Comment:
This multiplication was done earlier? Why recalculate?

Reply:
It was done earlier with `stub_r` but never `stub_z`